### DISCUSSION TITLE
FreeRTOS + SMP support in OpenOCD

### DISCUSSION THREAD

**Discussion by maxgerhardt** [earlephilhower/arduino-pico#1122] [Q&A]
With [the current `rp2040.cfg`](https://github.com/raspberrypi/openocd/blob/rp2040/tcl/target/rp2040.cfg), both cores are added with `-rtos hwthread` to make them appear as individual hardware threads when GDB connects to the GDB server opened by GDB and `info threads` is executed.

E.g. with the test firmware described in https://community.platformio.org/t/raspberry-picoprobe-openocd-freertos-debugging/31508/6?u=maxgerhardt

```
C:\Users\Max\.platformio\packages\toolchain-rp2040-earlephilhower\bin\arm-none-eabi-gdb.exe C:\Users\Max\temp\pico_first\.pio\build\rpipico\firmware.elf  -ex "target extended-remote :3333"
[...]
(gdb) info threads
  Id   Target Id                                       Frame
* 1    Thread 1 (Name: rp2040.core0, state: undefined) vApplicationMinimalIdleHook () at C:\Users\Max\.platformi
   o\packages\framework-arduinopico\libraries\FreeRTOS\src\variantHooks.cpp:184
  2    Thread 2 (Name: rp2040.core1, state: undefined) 0x00000000 in ?? ()
  ```
However, how do we make OpenOCD recognize that we have symmetric multiprocessing **and** FreeRTOS enabled with the FreeRTOS version used in Arduino-Pico?  Like in the ESP32 version, where we can see all FreeRTOS tasks that each CPU is executing?

![grafik](https://user-images.githubusercontent.com/26485477/212558917-e33a70cb-2f47-42a0-b094-f3437601d7a3.png)

OpenOCD has the `-rtos FreeRTOS` option, however, that then conflicts with `-rtos hwthread`.

As an experimentation for an easier case, only looking at Core0 running FreeRTOS, I have commented out the 

```
target smp $_TARGETNAME_0 $_TARGETNAME_1
```
line with a `#` and changed `-rtos hwthread` to `-rtos FreeRTOS`.

However, when attaching GDB then, OpenOCD complains.
```
Warn : RTOS FreeRTOS not detected. (GDB could not find symbol 'pxCurrentTCB')
```
and indeed GDB doesn't see any FreeRTOS tasks (there should be at least the idle task, timer task and a own-created "blink" task, and loop task)

```
(gdb) info threads
  Id   Target Id         Frame
* 1    Remote target     vApplicationMinimalIdleHook () at C:\Users\Max\.platformio\packages\framework-arduinopico\libraries\FreeRTOS\src\vari
   antHooks.cpp:184
```
And indeed when I ask GDB to print the symbol pxCurrentTCB, it says no.

```
(gdb) p pxCurrentTCB
No symbol "pxCurrentTCB" in current context.
```

looking at the FreeRTOS code however, it seems to [have deviated from mainline FreeRTOS](https://github.com/earlephilhower/FreeRTOS-Kernel/blob/b5942ddd071f426ff256b434bbcbbe750dfc8cfe/tasks.c#L329-L330)

```cpp
PRIVILEGED_DATA TCB_t * volatile pxCurrentTCBs[ configNUM_CORES ] = { NULL };
#define pxCurrentTCB    xTaskGetCurrentTaskHandle()
```
```cpp
xReturn = pxCurrentTCBs[ portGET_CORE_ID() ];
```

and these symbols *do* exist
```cpp
(gdb) p xTaskGetCurrentTaskHandle
$1 = {TaskHandle_t (void)} 0x10003778 <xTaskGetCurrentTaskHandle>
(gdb) p pxCurrentTCBs
$2 = {0x200034e4 <xIdleTCBBuffers.2>, 0x20001024 <vApplicationGetIdleTaskMemory::xIdleTaskTCB>}
```

The ESP32-Openocd seems to have some mods for this to recognized pxCurrentTCBs?

https://github.com/espressif/openocd-esp32/blob/master/src/rtos/FreeRTOS.c#L283-L301

Question is, how do we get FreeRTOS-awareness and SMP awareness into OpenOCD or activate it.

**Comment by maxgerhardt** [comment]
Issue also opened at https://github.com/raspberrypi/openocd/issues/73.

**Reply by earlephilhower** [reply]
The `TCB`->`TCBs` is part of the normal `SMP` branch of FreeRTOS, not something unique here, so it seems like this is a problem for *all* SMP FreeRTOS implementations.

I did do some changes to implement Newlib REENT swapping (which upstream did around the same time as my own changes), on the intercore FIFO handling to allow flash writes, and some minor mutexing to work w/the core code.  But nothing in the actual scheduling logic...

(I just wish the RPi team would submit their Picoprobe changes upstream so we wouldn't have to track multiple repos... :disappointed: )

**Reply by maxgerhardt** [reply]
>The TCB->TCBs is part of the normal SMP branch of FreeRTOS, not something unique here, so it seems like this is a problem for all SMP FreeRTOS implementations.

Yes that's something I'm worried about, because looking at mainline latest openocd, I don't see any checks to `pxCurrentTCBs`. So if FreeRTOS does have a standardized SMP branch with these changed symbols, but even current OpenOCD doesn't know about this, it's difficult.

https://github.com/openocd-org/openocd/blob/master/src/rtos/FreeRTOS.c#L113-L127
